oslabs-beta/apollo11
# Use actionHookForDevTools to listen for GraphQL operations

**Feature**
Detect GraphQL operations

**Description**
This PR changes how we detect GraphQL operations.  Previously, we relied on network events but this left a gap in detection of operations that did not go across the network.  We will now use the `actionHookForDevTools` function to inject our callback.  This hook will invoke our callback every time the Apollo Client makes a query or mutation, even if is done locally and not across the network.  When invoked, the callback will grab the necessary Apollo Client data and send it our app for processing in the Event Log.

There are several key things to note here:
- There are three counters in v3 of the Apollo Client `queryManager`:  `requestIdCounter`, `queryIdCounter`, `mutationIdCounter`.  We rely on these counters to know which operation was just made, so that we know whether to inspect the `queriesStore` or the `mutationStore` to get the actual GraphQL operation data (`operationName`, `query` or `mutation`, `variables`, etc).
- The callback is fired even when there is no new operation, but when the "state" of the operation has changed from `loading` to `complete`.  This will signal when to get the cache and update the Event Log appropriately.  We use a `lastEventId` property to record the last operation that just occurred, so that we can update the cache for that event again.
- We now store the network-based operations into its own Network Events log.  This Network Events log also no longer needs the cache so it no longer sends a message to the `contentScript` to get it.
- There isn't any protection to ensure that our callback is not overwritten by another extension (e.g. Apollo Dev Tools).  Some work will need to be done to ensure that we re-inject our callback as needed.

- There is only one counter in v2 of the Apollo Client:  `idCounter`.  Currently we don't have any way of inspecting which operation just occurred, so for the time being these events are ignored (i.e. no v2 support).  This is done because of the way we render the Event Log:  a chronological timeline of GraphQL operations.  
- We could add support for v2 if we split the Event Log between queries and mutations, which is what the Apollo Dev Tools does, since there is no mixing of operations and it simply displays the content of their respective stores.

**How to test**
- Load the extension and navigate to a test website that has a mix of local vs network events for GraphQL operations
- Observe that the Event Log is populated with entries even when there are no GraphQL network events
- Observe that the Network Log is populated with only network-based GraphQL events
